pkg/validation/internal/operatorhub.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/api/pkg/validation/errors"
+	interfaces "github.com/operator-framework/api/pkg/validation/interfaces"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var OperatorHubValidator interfaces.Validator = interfaces.ValidatorFunc(validateOperatorHub)
+
+func validateOperatorHub(objs ...interface{}) (results []errors.ManifestResult) {
+	for _, obj := range objs {
+		switch v := obj.(type) {
+		case *manifests.Bundle:
+			results = append(results, validateBundleOperatorHub(v))
+		}
+	}
+	return results
+}
+
+func validateBundleOperatorHub(bundle *manifests.Bundle) errors.ManifestResult {
+	result := errors.ManifestResult{Name: bundle.Name}
+
+	if bundle == nil {
+		result.Add(errors.ErrInvalidBundle("Bundle is nil", nil))
+		return result
+	}
+
+	if bundle.CSV == nil {
+		result.Add(errors.ErrInvalidBundle("Bundle csv is nil", bundle.Name))
+	}
+
+	errs := validateHubCSVSpec(*bundle.CSV)
+	for _, err := range errs {
+		result.Add(errors.ErrInvalidCSV(fmt.Sprintf("%s", err), bundle.CSV.GetName()))
+	}
+
+	return result
+}
+
+func validateHubCSVSpec(csv v1alpha1.ClusterServiceVersion) []error {
+	var errs []error
+
+	errs = append(errs, checkAlmExamples(csv)...)
+
+	if csv.Spec.Provider.Name == "" {
+		errs = append(errs, fmt.Errorf("csv.Spec.Provider.Name not specified"))
+	}
+
+	for _, maintainer := range csv.Spec.Maintainers {
+		if maintainer.Name == "" || maintainer.Email == "" {
+			errs = append(errs, fmt.Errorf("csv.Spec.Maintainers elements should contain both name and email"))
+		}
+		if maintainer.Email != "" {
+			_, err := mail.ParseAddress(maintainer.Email)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("csv.Spec.Maintainers email %s invalid", maintainer.Email))
+			}
+		}
+	}
+
+	for _, link := range csv.Spec.Links {
+		if link.Name == "" || link.URL == "" {
+			errs = append(errs, fmt.Errorf("csv.Spec.Links elements should contain both name and url"))
+		}
+		if link.URL != "" {
+			_, err := url.ParseRequestURI(link.URL)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("csv.Spec.Links url %s invalid", link.URL))
+			}
+		}
+	}
+
+	if capability, ok := csv.ObjectMeta.Annotations["capabilities"]; ok {
+		validCapabilities := map[string]struct{}{
+			"Basic Install":     struct{}{},
+			"Seamless Upgrades": struct{}{},
+			"Full Lifecycle":    struct{}{},
+			"Deep Insights":     struct{}{},
+			"Auto Pilot":        struct{}{},
+		}
+
+		if _, ok := validCapabilities[capability]; !ok {
+			errs = append(errs, fmt.Errorf("csv.Metadata.Annotations.Capabilities %s is not a valid capabilities level", capability))
+		}
+	}
+
+	if csv.Spec.Icon != nil {
+		// only one icon is allowed
+		if len(csv.Spec.Icon) != 1 {
+			errs = append(errs, fmt.Errorf("csv.Spec.Icon should only have one element"))
+		}
+
+		for _, icon := range csv.Spec.Icon {
```

Comment:
Should all icons be validated if there should be exactly one? Was this intended to be a no-op if `len(csv.Spec.Icon) == 0`?